pkg/publisher/combo/fanout.go

```diff
@@ -102,13 +116,49 @@ func (f *fanoutPublisher) PublishShardResult(
 		return p.PublishShardResult(ctx, shard, hostID, shardResultPath)
 	})
 
-	// Just return the first storage spec that we get
-	select {
-	case value := <-valueChannel:
-		return value, nil
-	case err := <-errorChannel:
-		return model.StorageSpec{}, err
+	type result struct {
+		Sender publisher.Publisher
+		Value  model.StorageSpec
 	}
+
+	timeoutChannel := make(chan bool, 1)
+	results := make([]result, 0)
+
+loop:
+	for {
+		select {
+		case value := <-valueChannel:
+			results = append(results, value)
+			if len(results) == len(f.publishers) {
+				// break because everyone returned
+				break loop
+			}
+
+			// start timeout for other results when first result is returned
+			if len(results) == 1 {
+				go func() {
+					// TODO: Make timeout value configurable
+					time.Sleep(time.Second * 2)
+					timeoutChannel <- true
+				}()
+			}
+
+		case <-timeoutChannel:
+			break loop
+		case err := <-errorChannel:
+			return model.StorageSpec{}, err
+		}
+	}
+
+	// search for preferred result
+	for _, result := range results {
+		if result.Sender == f.preferred {
+			return result.Value, nil
+		}
+	}
+
+	// return first result if preferred not found
+	return results[0].Value, nil
```

Comment:
Would a `map[Publisher]StorageSpec` make more sense here? And then for the default case just use `maps.Values`?

Reply:
Yup, makes more sense 👍🏻